New optional feature - display out of stock products last

gadnis opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR adds a button in BO-> Shop Parameters -> Product Settings -> Pagination TAB, an option to display out of stock products last. If this feature enabled, than products goes in this order: 1. inStock, 2. OutOfStock & Available for order, 3. OutOfStock. Affected pages: category, best-sales, new-products, price-drop. It doesn't rearrange products if FO user sorts by: name or sales, cause it would be not logical. A lot of people has thousands of products in database and thousands of them is with quantity 0 but allowed to order. Before this feature, these "available  to order" products went to the end of product list, with this feature it will go after the "available" products and out-of-stock products will be the last products.
| Type?         | new feature
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | "Fixes #9956". Or atleast one issue mentioned in comments.
| How to test?  | Turn off module: ps_facetedsearch (cause it overrides native functions for category pages). Go to: BO-> Shop Parameters -> Product Settings -> Pagination TAB. Switch "Display out-of-stock products last" to ON. Go to FO pages: any category, new-products, prices-drop, best-sales. Choose to sort by price, data added. You will see products sorted in the chunks: 1. inStock, 2. OutOfStock & Available for order, 3. OutOfStock.

Here is some pictures to help you understand what PR does:

BO: feature switch added:
![screen shot 2019-02-05 at 22 58 29](https://user-images.githubusercontent.com/10948341/52303798-9949da80-2999-11e9-9654-cef6643e6ce3.png)

FO: sort by Price examples:
In category pages:
![screen shot 2019-02-05 at 22 59 42](https://user-images.githubusercontent.com/10948341/52303853-cdbd9680-2999-11e9-881b-6923e3391cc5.png)

In best-sales page:
![screen shot 2019-02-05 at 23 02 07](https://user-images.githubusercontent.com/10948341/52303948-25f49880-299a-11e9-8d1b-d954c052a1fb.png)

In prices-drop page:
![screen shot 2019-02-05 at 23 02 54](https://user-images.githubusercontent.com/10948341/52303972-36a50e80-299a-11e9-8133-8d03b8e924f9.png)

In new-product page:
![screen shot 2019-02-05 at 23 01 26](https://user-images.githubusercontent.com/10948341/52303906-01002580-299a-11e9-8471-95ce0ebe50c2.png)

Sort by Date added in new-product page:
![screen shot 2019-02-05 at 23 03 47](https://user-images.githubusercontent.com/10948341/52304060-6c49f780-299a-11e9-886c-a619884cbdb9.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12441)
<!-- Reviewable:end -->

gadnis committed:
Add stock quantity variable inside cart summary notification error

When user adjust quantity in cart summary and wanted quantity is more than 'in-stock', than error notification shown, but with this improvement user knows maximum product quantity he can buy.

gadnis committed:
Update CartController.php

gadnis committed:
BO: Product settings- field added. FO: Products is sorting in price-drop, category pages, more to come

gadnis committed:
FO: sort inStock,OutOfStock & OOPS, OutOfStock.Pages:category,new-products,product-sales,prices-drop

gadnis committed:
removed unnecessary comment

gadnis committed:
changed ps.label_with_help for display_out_of_stock_last field

gadnis committed:
removed row string

matks commented:
@colinegin This is a new feature, what do you think 😉 ?

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_pagination.html.twig:
```diff
@@ -47,6 +47,11 @@
                     {{ form_widget(paginationForm.default_order_way) }}
                 </div>
+                <div class="form-group">
+                    {{ ps.label_with_help(('Display out-of-stock products last'|trans), ('If enabled, products will be in this order: 1. in-stock, 2. available for order but quantity is 0, 3. out-of-stock. This does not apply when ordering by name & sales'|trans({}, 'Admin.Shopparameters.Help'))) }}
```
Hello, can you explain the "ordering by name & sales" please? Also, I would write _If enabled, products will be displayed in this order: 1. in-stock, 2. available for order but quantity is 0, 3. out-of-stock. Note that it does not apply when ordering by name or sales._ instead, what do you think?

marionf commented:
Hello @gadnis 

In BO > Shop parameters > Products settings, you can already choose to filter by quantity by default

![capture d'écran_1844](https://user-images.githubusercontent.com/13449658/61857924-5d61ed80-aec5-11e9-84ca-c4b61e936e1b.png)

This way, when your customer arrives on a category page, "Relevance" is selected and your products are sorted by quantity.

The problem is that this option only takes into account the quantity and not if out of stock products can be ordered or not.

So, we would like to take this into account when quantity is selected as default order by.
Without adding a new option in the drop-down.

What dou think about that ?

gadnis commented:
Hi @marionf ! It's kind of good idea but not really.. If people would choose to order by name, I understand that they want to see product in A-Z or Z-A order. But when day choose to order by price, they aspect to see lets say cheapest products first and hope that the seller has those products in stock, cause why else order by price.. unless to spy or investigate product prices, which is not that offen the case to order by price. I Would add a switch (ON, OFF option) to show out-of-stock products last.

marionf commented:
Thank you for your answer @gadnis 

I would like to test your PR and maybe suggest some improvements but there are some conflicting files and this generates errors
Could you please resolve conflicts ?

gadnis commented:
Ok, @marionf, I will try to do that in the next week ;)

gadnis committed:
Latest 1.7.6.2 compatible change

Adding switch to enable or disable the feature

gadnis committed:
PS 1.7.6.2 compatible change 

Adding SwitchType to a Form. To be able switch off|on this feature

gadnis committed:
PS 1.7.6.2 compatible change

Adding configuration field 'display_out_of_stock_last'. It will be used to either enable or disable the feature

gadnis committed:
PS 1.7.6.2 compatible chage

This one is not related to the feature. I'm not sure why it is in this PR. But it is useful.

gadnis committed:
PS 1.7.6.2 compatible change

Adding two methods: orderByOOSP and sqlOrderByOOSP
There is All "facets" where we allow to sort out-of-stock last in method sqlOrderByOOSP 
And method orderByOOSP helps to do that later

gadnis committed:
PS 1.7.6.2 compatible change

gadnis committed:
PS 1.7.6.2 compatible change

changes marked with comment: ORDER BY out-of-stock products last

gadnis committed:
PS 1.7.6.2 compatible change

Allowing order products as described in PR
Changes marked with comment: ORDER BY out-of-stock products last

gadnis reviewed (commented):
Code is now compatible with PS 1.7.6.2

gadnis committed:
Merge branch 'develop' into new-feature-display-out-of-stock-products-last

gadnis commented:
> Thank you for your answer @gadnis
> 
> I would like to test your PR and maybe suggest some improvements but there are some conflicting files and this generates errors
> Could you please resolve conflicts ?

Hi, @marionf , finally I updated our shop to the latest, and now I had to put back my code there. So I change the code in this PR also. Please review and merge it 😁 

Now I will go to find out how to do the same feature on ps_facetedsearch module, cause it was refactored and my code is no fit in there...

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_pagination.html.twig:
```diff
@@ -55,7 +55,13 @@
         </div>
-
+        <div class="form-group row">
+          {{ ps.label_with_help(('Display out-of-stock products last'|trans), ('If enabled, products will be displayed in this order: 1. in-stock, 2. available for order but quantity is 0, 3. out-of-stock. Note that it apply only when ordering by quantity, price or dates'|trans({}, 'Admin.Shopparameters.Help'))) }}
```
I would not use the word 'reordering' here, it can be ambiguous. What about _Note that it only applies when filtering by quantity, price or date._ instead?

gadnis committed:
Note changed

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_pagination.html.twig:
```diff
@@ -55,7 +55,13 @@
         </div>
-
+        <div class="form-group row">
+          {{ ps.label_with_help(('Display out-of-stock products last'|trans), ('If enabled, products will be displayed in this order: 1. in-stock, 2. available for order but quantity is 0, 3. out-of-stock. Note that it only applies when filtering by quantity, price or date'|trans({}, 'Admin.Shopparameters.Help'))) }}
```
```suggestion
          {{ ps.label_with_help(('Display out-of-stock products last'|trans), ('If enabled, products will be displayed in this order: 1. in-stock, 2. available for order but quantity is 0, 3. out-of-stock. Note that it only applies when filtering by quantity, price or date.'|trans({}, 'Admin.Shopparameters.Help'))) }}
```

gadnis committed:
Update src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_pagination.html.twig

Co-Authored-By: LouiseBonnard <32565890+LouiseBonnard@users.noreply.github.com>

PierreRambaud commented:
Hi @gadnis ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!
